### ISSUE TITLE
GenericWhitespace false positive in google_checks.xml with generic method call

### ISSUE THREAD

**Issue by ctubbsii** [checkstyle/checkstyle#533]
A generic method is called with an ambiguously typed parameter:
`Collections.<Long> emptyList()`

And results in "GenericWhitespace: GenericWhitespace '>' is followed by whitespace."

But, that should be okay. Google style guide does not specify regarding whitespace here.

**Comment by romani** [comment]
please review http://checkstyle.sourceforge.net/config_whitespace.html#GenericWhitespace

here is example "boolean same = Util.<Integer, String>compare(p1, p2);  // Generic preceded method name"

We treat generic as part of method, and space is not do any good there to our mind.

If you disagree , please provide detailed explanation, please provide examples from well known open-source projects, the more examples the better.

**Comment by ctubbsii** [comment]
I don't disagree that it should be considered part of the method, and space doesn't make sense, but 1) it's undefined in the google styleguide, so should not be enforced in the google checks, and 2) the default formatting in Eclipse includes the space.

**Comment by romani** [comment]
> but 1) it's undefined in the google styleguide, so should not be enforced in the google checks

not exactly defined and do skip - it is not good approach.
we did that decision base on testing over Guava, only two places in whole Guava violate that 
http://checkstyle.sourceforge.net/reports/google-style/guava/xref/com/google/common/collect/testing/google/MultimapTestSuiteBuilder.html#L79
http://checkstyle.sourceforge.net/reports/google-style/guava/xref/com/google/common/collect/testing/google/MultimapTestSuiteBuilder.html#L266
So it mean for us, that Google is following rules that might be the same in Checkstyle.
Please provide us evidences that " " should be allowed, evidence = links to code.

FYI:
http://checkstyle.sourceforge.net/google_style.html
http://checkstyle.sourceforge.net/reports/google-style/guava/
https://github.com/google/guava/issues/1891 (here Google agreed to change Style guide to be more exact on what violation should be shown, but there are nothing about your issue )


> 2) the default formatting in Eclipse includes the space.

default and correct is different terms. Eclipse do too much for default that checkstyle does not like.
Eclipse config should be changed, of checkstyle suppressed - it is up to a user.



Does it make sense ?

**Comment by ctubbsii** [comment]
You seem to be arguing that it's *okay* that the checkstyle enforcement of google style include enforcement of things not defined in the google style. Thus, you are admitting that this is not enforcing the google style, but rather a modified, more strict version of google style. I bring this up, because the Google Style Guide does not define this behavior, and *their own Eclipse formatter* produces such code. I have not been able to figure out a way to eliminate this space in Eclipse.

I'm not sure why you want links to code... all that will show is that the code doesn't comply with this enforcement. It will not help determine which rules should be enforced and which shouldn't be. That should be determined by the Google Style Guide's explicit wording. Anything not strictly defined, should be permitted (unless it violates JLS recommendations or some such reasonable minimums).

If you think this *should* be a rule in the Google Style Guide enforcement, then maybe you should open a ticket with that, so they can update their guide with strictly-defined behavior which can be enforced? Otherwise, why enforce rules not established, and still call it google checks? It's not. It diverges from google. As you've already demonstrated with those two violations in Guava... this is not a rule which should be enforced, because it is *permitted* (currently).

**Comment by romani** [comment]
> I have not been able to figure out a way to eliminate this space in Eclipse.

in Eclipse preferences: Java > Code Style > formatter 

![screenshot-generic-formatter1](https://cloud.githubusercontent.com/assets/812984/5842401/5a1f5f64-a159-11e4-9546-2959b94d8e48.png)
![screenshot-generic-formatter2](https://cloud.githubusercontent.com/assets/812984/5842402/5a217934-a159-11e4-9bf7-833db9455e46.png)

So that behavior is configurable in Eclipse.

> and their own Eclipse formatter produces such code

please provide link to Google formatter, I need to recheck, I wil not be suprised if Google do not use it :) .

> Anything not strictly defined, should be permitted

I am kind of agree with you, but after spending 3 month in Google style and Guava code , I do not agree.

> If you think this should be a rule in the Google Style Guide enforcement, then maybe you should open a ticket with that,

We are try to cooperate with Google to refine Style guideline, Google already agreed to change style, as a lot is missed there and need to be more exact.

> As you've already demonstrated with those two violations in Guava... this is not a rule which should be enforced, because it is permitted (currently).

Code in Guava is not vollowing any guidelines, so 2 violations is ok, most likely with space enforced we will have more violations. I donot remember exactly that case, but I remember that we did investigation of what is more rule for that project. Am ok to remove that enforcement but lets have official confirmation.

**Comment by romani** [comment]
https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s4.6.2-horizontal-whitespace

> Beyond where required by the language or other style rules, and apart from literals, comments and Javadoc, a single ASCII space also appears in the following places only.

attention to last phrase - "a single ASCII space also appears in the following places only". last word "only" !!!. 

Your case is not defined in following examples in guide, so not allowed.

**Comment by ctubbsii** [comment]
Excellent! I had overlooked that bit of the guide! Thank you for your time and assistance. And, now that I'm double-checking their formatter, it seems they do have this option set...
(https://code.google.com/p/google-styleguide/source/browse/trunk/eclipse-java-google-style.xml#66)

So, I don't know which formatter I tested this with which didn't work... but I apologize for (what seems to be) a complete waste of both our time. Sorry for that.
